drivers/d3d12/rendering_device_driver_d3d12.cpp

```diff
@@ -631,6 +633,33 @@ void RenderingDeviceDriverD3D12::_resource_transitions_flush(ID3D12GraphicsComma
 		// to avoid attempting the single-barrier optimization.
 		static const uint32_t SINGLE_BARRIER_ATTEMPT_MAX_NUM_SUBRESOURCES = 48;
 
+		uint32_t first_subres_to_init = UINT32_MAX;
+		uint32_t num_subres_to_init = 0;
+		auto _add_subresource_to_initialize = [&](uint32_t p_subresource) {
+			if (first_subres_to_init == UINT32_MAX) {
+				first_subres_to_init = p_subresource;
+				num_subres_to_init = 1;
+			} else if (p_subresource == first_subres_to_init + num_subres_to_init) {
+				num_subres_to_init++;
+			} else {
+				D3D12_DISCARD_REGION dr = {};
+				dr.FirstSubresource = first_subres_to_init;
+				dr.NumSubresources = num_subres_to_init;
+				p_cmd_list->DiscardResource(br.dx_resource, &dr);
+				first_subres_to_init = p_subresource;
+				num_subres_to_init = 1;
+			}
+			res_states->subresource_states[p_subresource].init_state = ResourceInfo::States::ResourceStateEx::INIT_STATE_GARBAGE;
+		};
+		auto _flush_initialize_set = [&]() {
```

Comment:
This is only called once so I'd avoid a lambda

Reply:
My idea was to keep those very related pieces of functionality together so they can be studied more easily.

Reply:
It looks like the call locations are over 100 lines of code away though. So I would also prefer elevating these to proper functions

Reply:
Those changes belong to the base PR included here (#87872). I've amended it with an approach that isn't either lambdas nor member functions. I believe the one one has all the advantages. I've rebased this PR on top of the amended one.

Reply:
The new approach doesn't do any of that anymore.